### PR TITLE
[RFC] pinctrl: stm32: uart low power pinmux (DNM)

### PR DESCRIPTION
**Pull request by erwango**
These commits shows how one would add a low power alternate pin configuration to a driver.

Alternate low power pinmux is put directly manualy in board files, but alternatively it could aslo be generated and stored in *-pinctrl.dtsi under hal_stm32.

Note: This change is only meant for support on pinctrl API discussion. It is not done to be merged

### CODE REVIEW

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -1444,6 +1446,23 @@ static int uart_stm32_set_power_state(const struct device *dev,
...
+		err = stm32_dt_pinctrl_configure(config->lp_pinctrl_list,
+						 config->lp_pinctrl_list_size,
+						 (uint32_t)UART_STRUCT(dev));
+		if (err < 0) {
+			return err;
+		}
+
+	} else {
+		/* Configure dt provided device signals when available */
+		err = stm32_dt_pinctrl_configure(config->pinctrl_list,
```
Comment (alexanderwachter) [inline drivers/serial/uart_stm32.c:1460]:
```suggestion
		err = stm32_dt_pinctrl_configure(config->lp_pinctrl_list,
```

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_stm32.c (hunk 1 above)
Comment (erwango) [reply drivers/serial/uart_stm32.c:1460]: At this point we need to provide active state pinmux configuration.
So I think this is correct.
I'll update comments though

Comment (erwango) [reply drivers/serial/uart_stm32.c:1457]: @alexanderwachter Can you review ?

**Review by alexanderwachter — APPROVED** [review]

File: boards/arm/stm32l562e_dk/stm32l562e_dk.dts [hunk 2]
```diff
@@ -31,8 +31,18 @@
...
 };
 
+&pinctrl {
+	usart1_lp_tx_pa9: usart1_lp_tx_pa9 {
+		pinmux = <STM32_PINMUX('A', 9, ANALOG)>;
+	};
+	usart1_lp_rx_pa10: usart1_lp_rx_pa10 {
+		pinmux = <STM32_PINMUX('A', 10, ANALOG)>;
+	};
+};
```
Comment (gmarull) [inline boards/arm/stm32l562e_dk/stm32l562e_dk.dts:41]: lp variants should be provided by the pinctrl files, ref. https://github.com/zephyrproject-rtos/hal_stm32/blob/master/scripts/genpinctrl/stm32-pinctrl-config.yaml#L30

File: boards/arm/stm32l562e_dk/stm32l562e_dk.dts [hunk 3]
```diff
@@ -31,8 +31,18 @@
...
+
 &usart1 {
 	pinctrl-0 = <&usart1_tx_pa9 &usart1_rx_pa10>;
+	pinctrl-1 = <&usart1_lp_tx_pa9 &usart1_lp_rx_pa10>;
```
Comment (gmarull) [inline boards/arm/stm32l562e_dk/stm32l562e_dk.dts:45]: I thik that we should introduce the `pinctrl-names` option as in Linux, i.e.
```devicetree
pinctrl-names = "default", "sleep";
```
C Macros used to obtain pinctrl information should refer to states and not indexes. We may need some devicetree scripts changes to handle that.

File: boards/arm/stm32l562e_dk/stm32l562e_dk.dts (hunk 2 above)
Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk.dts:41]: Sure, that is my current though as well, but aim was to provide material for discussion, this PR is not meant for merge (cf PR description).

File: boards/arm/stm32l562e_dk/stm32l562e_dk.dts (hunk 3 above)
Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk.dts:45]: Agreed. I don't think we need more device tree macros for that (at least from devicetree.h).
Everything should be there, what is missing is the `ST_STM32_DT_INST_PINCTRL_FOO` version.

Comment (mbolivar-nordic) [reply boards/arm/stm32l562e_dk/stm32l562e_dk.dts:45]: Let me know if you do decide you need something.

Comment (gmarull) [reply boards/arm/stm32l562e_dk/stm32l562e_dk.dts:45]: Do we have the infrastructure to query by name on entries indexed like pinctrl, i.e. `pinctrl-0, pinctrl-1...pinctrl-N`? Can we get now the index `N` of an entry given its name, to then get the contents of `pinctrl-N`? I also think we need to support `pinctrl-N` and `pinctrl-names` in a generic way (we now repeat that on every binding).

**Pull request closed without merging**